feat(test runner): server side mocking (#34520)
microsoft/playwright: docs/src/mock.md

```

For more details, see [WebSocketRoute].

## Mock Server
* langs: js

By default, Playwright only has access to the network traffic made by the browser.
To mock and intercept traffic made by the application server, use Playwright's **experimental** mocking proxy. Note this feature is **experimental** and subject to change.

The mocking proxy is a HTTP proxy server that's connected to the currently running test.
If you send it a request, it will apply the network routes configured via `page.route` and `context.route`, reusing your existing browser routes.

To get started, enable the `mockingProxy` option in your Playwright config:

```ts
export default defineConfig({
  use: { mockingProxy: true }
});
```

> Playwright will now inject the proxy URL into all browser requests under the `x-playwright-proxy` header.

Reviewer: Let's have a list of limitations somewhere that would mention:
- extra header will affect CORS options preflight;
- requests on the server that were not made in response to the browser request will not be routed;
- etc.

On your server, read the URL in this header and prepend it to all outgoing traffic you want to intercept:

```js
const headers = getCurrentRequestHeaders(); // this looks different for each application
const proxyURL = decodeURIComponent(headers.get('x-playwright-proxy') ?? '');
await fetch(proxyURL + 'https://api.example.com/users');
```

> Prepending the URL will direct the request through the proxy. You can now intercept it with `context.route` and `page.route`, just like browser requests:

Reviewer: ```suggestion
Prepending the URL will direct the request through the proxy. You can now intercept it with [`method: BrowserContext.route`] and [`method: Page.route`], just like browser requests:
```

> ```ts

Reviewer: Please mark all snippets as "js". That's how we handle it in docs for historical reasons.

```suggestion
```js
```

// shopping-cart.spec.ts
import { test, expect } from '@playwright/test';

test('checkout applies customer loyalty bonus points', async ({ page }) => {
  await page.route('https://users.internal.example.com/loyalty/balance*', (route, request) => {
    await route.fulfill({ json: { userId: 'jane@doe.com', balance: 100 } });
  });

  await page.goto('http://localhost:3000/checkout');

  await expect(page.getByRole('list')).toMatchAriaSnapshot(`
    - list "Cart":
      - listitem: Super Duper Hammer
      - listitem: Nails
      - listitem: 16mm Birch Plywood
    - text: "Price after applying 10$ loyalty discount: 79.99$"
    - button "Buy now"
  `);
});
```

Now, prepending the proxy URL manually can be cumbersome. If your HTTP client supports it, consider updating your client baseURL ...

```js
import { axios } from 'axios';

const api = axios.create({

>   baseURL: proxyURL + 'https://jsonplaceholder.typicode.com',

Reviewer: Is that going to work? I thought we are only shipping the version with `proxyURL` per each browser request, so initializing it once will not work.

});
```

... or setting up a global interceptor:

```js
import { axios } from 'axios';

axios.interceptors.request.use(async config => {

>   config.baseURL = proxyURL + (config.baseURL ?? '/');

Reviewer: Please include the snippet that gets current `proxyURL` here. Otherwise it's unclear what's happening.

  return config;
});
```

```js
import { setGlobalDispatcher, getGlobalDispatcher } from 'undici';

const proxyingDispatcher = getGlobalDispatcher().compose(dispatch => (opts, handler) => {
  opts.path = opts.origin + opts.path;

>   opts.origin = proxyURL;

Reviewer: Same here.

  return dispatch(opts, handler);
});
setGlobalDispatcher(proxyingDispatcher); // this will also apply to global fetch
```

:::note

> Note that this style of proxying, where the proxy URL is prended to the request URL, does *not* use [`CONNECT`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT), which is the common way of establishing a proxy connection.

Reviewer: ```suggestion
Note that this style of proxying, where the proxy URL is prepended to the request URL, does *not* use [`CONNECT`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT), which is the common way of establishing a proxy connection.
```
